pip/PIP-334.md

```diff
@@ -0,0 +1,201 @@
+        if (!nonOrphanedLedgers.contains(ledgerId)) {
+          try {
+            semaphore.acquire();
+            openLedger(ledgerId).thenAccept((infosx) -> {
+              if (!infosx.isPulsarLedger()) {
+                log.warn("Ledger:{} is not a pulsar ledger.", ledgerId);
+                semaphore.release();
+                return;
+              }
+
+              TopicName topic = infosx.getLedgerTopic();
+              if (topic == null) {
+                maybeOrphanedLedgers.add(infosx);
+                semaphore.release();
+              } else {
+                infosx.stillExistsInMetaStore().thenAccept((exists) -> {
+                  if (!exists) {
+                    processOrphanedLedger(infosx, staleTime).whenComplete((r, ex) -> {
+                      r.ifPresent(orphanedLedgers::add);
+                      semaphore.release();
+                    });
+                  } else {
+                    getTopicUsedLedgersAsync(topic.toString()).thenAccept((ledgers) -> {
+                      ledgers.forEach((lid) -> {
+                        nonOrphanedLedgers.add(lid);
+                        maybeOrphanedLedgers.remove(lid);
+                      });
+                      if (!ledgers.contains(ledgerId)) {
+                        processOrphanedLedger(infosx, staleTime).whenComplete((r, ex) -> {
+                          r.ifPresent(orphanedLedgers::add);
+                          semaphore.release();
+                        });
+                      } else {
+                        semaphore.release();
+                      }
+
+                    }).exceptionally((ex) -> {
+                      if (ex.getCause() instanceof PulsarAdminException.NotFoundException) {
+                        if (infosx.isSchemaLedger()) {
+                          maybeOrphanedLedgers.add(infosx);
+                          semaphore.release();
+                        } else {
+                          processOrphanedLedger(infosx, staleTime).whenComplete((r, ex1) -> {
+                            r.ifPresent(orphanedLedgers::add);
+                            semaphore.release();
+                          });
+                        }
+                      } else {
+                        log.error("Failed to get topic: {} ledger list: {}", topic, ex);
+                        semaphore.release();
+                      }
+
+                      return null;
+                    });
+                  }
+
+                }).exceptionally((ex) -> {
+                  log.error("Failed to get topic: {} ledger list: {}", topic, ex);
+                  semaphore.release();
+                  return null;
+                });
+              }
+
+
+            });
+          } catch (InterruptedException e) {
+            asyncResponse.resume(new RestException(e));
+          }
+        }
+      });
+
+      while (semaphore.availablePermits() != semaphorePermits) {
```

Comment:
Is it better to use CountDownLatch here?

Reply:
The purpose of using `semaphore` here is to limit concurrent access to metadata, so using `CountDownLatch` may not meet the requirements.